### ISSUE TITLE
Add matrix factorisations to linalg

### ISSUE THREAD

**Issue by karlnapf** [shogun-toolbox/shogun#2526]
We need
- Cholesky
- SVD
- QR

@lambday @khalednasr I leave this to you

**Comment by SunilMahendrakar** [comment]
@karlnapf  @lambday do I start working on this?

**Comment by karlnapf** [comment]
Go for it. @lambday will guide you.
First step: Linear solves
 * Direct solve (A*x =b ), similar to [numpy.lingalg.solve](http://docs.scipy.org/doc/numpy/reference/generated/numpy.linalg.solve.html)
 * Solve via cholesky, triangular solve, pseudo inverse, see also [here](http://docs.scipy.org/doc/scipy/reference/linalg.html)

Second step: Factorisations
 * The three mentioned above, plus eigendecomposition
 * Also offer solve methods given the factorisation

Third:
 * backends: eigen3, lapack, (native?? though not sure here)

**Comment by SunilMahendrakar** [comment]
@karlnapf Okay! :)

**Comment by yingryic** [comment]
hi, @SunilMahendrakar are you are still working on this problem. If you are interested in the deep learning project, I wonder if you could let me look into this. Thanks

**Comment by SunilMahendrakar** [comment]
@yingryic Hi, I was going to work on this but little bit slowly.. If you want to finish this you can proceed. You may refer to [eigen3 doc](http://eigen.tuxfamily.org/dox/classEigen_1_1LLT.html) to get started.

**Comment by yingryic** [comment]
@SunilMahendrakar Thanks, if you havent started, I guess I could get going.
After checking the linsolver directory, I guess shogun currently only has direct solver for sparse symmetric matrix. So are you interested in a solver of general dense matrix (non sysmmtric) based on LU factorization, or the least square solution using QR decompositon or SVD. (one class for each of the methods?)
All this can be implemented easily by using the eigen library or you need to use other libraries.
Tell me what you think @karlnapf @lambday

**Comment by karlnapf** [comment]
@yingryic 
Hi, cool that you are getting involved here.
Let's maybe start with a plain linear solve? I checked what numpy does:

http://docs.scipy.org/doc/numpy/reference/generated/numpy.linalg.solve.html
http://docs.scipy.org/doc/numpy/reference/generated/numpy.linalg.lstsq.html#numpy.linalg.lstsq

First wraps LAPACK's _gesv, i.e. assuming full rank and then solving exactly
Second does least squares.

If we had those, we can already replace a lot of things within Shogun (we can also justify a native implementation for them, there is a Gauss-Seidel implementation in KRR I think, you can copy that over if not some other class, search for it)

We need first to define the interface, and then provide all backend implementations. In particular useful will be to have viennacl here, see for example
http://viennacl.sourceforge.net/viennacl-examples-dense-matrix.html

Once this is done, we want solvers like solve_triangular, solve_cholesky, solve_qr, solve_svd which do the job based on factorisations that one computed before. (Also need methods for that first)
http://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.cho_solve.html
http://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.lu_solve.html#scipy.linalg.lu_solve
http://math.stackexchange.com/questions/109329/can-qr-decomposition-be-used-for-matrix-inversion


Keep in mind this linalg stuff is about *interface* first, and the providing implementations for the methods in various backends (native/eigen/lapack)

**Comment by sudk1896** [comment]
@karlnapf @OXPHOS : I see that (@OXPHOS) your summer work was with linalg. I also checked your repo and saw your WIP with regards to `Cholesky` Decomposition. I would like to get started with `QR Decompostion`, since no one else is working with it. Could you give me some pointers/general rules, guidelines etc on how I should go about adding it. Thanks.

**Comment by karlnapf** [comment]
@sudk1896 I don't think it is a good idea to get involved here. We will merge linalg soon and then there will be tons of stuff (easy) to do, but currently things a bit more complicated and not really good for getting involved in Shogun

**Comment by sudk1896** [comment]
@karlnapf : Okay, no harm done. I haven't looked into it anyway.

**Comment by OXPHOS** [comment]
Question: For LDLT, LU, QR and SVD, more than one matrix will be obtained after decomposition. It seems that the best way to fit in is like: `svd_factor(matrix, bool matrixU=true)`. Is this kinda weird?

**Comment by karlnapf** [comment]
SVD returns three objects, U, V, and a vector S with the singular values.
Need to think about the interface.

I guess there can be two interfaces

 * a solve method that just uses the specified  method and matrix to solve a linear system (no factorization returned)
 * a factor method that turns a matrix into multiple factors (different for every method)
 * a solve method that takes the factors (or just a few of them), and then behaves like the first solve method, but with precomputed factorizations
 * The helper methods (solve_triagnular, etc) should also be available in linalg btw. E.g. for Cholesky you use a traingular solve inside, but maybe somebody also wants to use this expicitly. Will be the same for qr
